---
title: "sigident - Howto Microarray"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{sigident-Howto_Microarray}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::knitr}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(sigident)
library(knitr)
```

# Prerequisites and Installation 

## Install sigident package

```{r eval=FALSE}
options('repos' = 'https://ftp.fau.de/cran/')
install.packages("devtools")
devtools::install_git("https://gitlab.miracum.org/clearly/sigident.git")
```

# Data Preparation 

In order to use this R package and its functions, you need to prepare a merged gene expression dataset. 
This example uses the GEO lung cancer studies "GSE18842", "GSE19804" and "GSE19188", which contain in total 367 samples (197 tumor; 170 non-tumor) and 54,675 transcripts gained by using Affymetrix GeneChip Human Genome U133 Plus 2.0 Array (platform GPL570).

## Download of GEO datasets and normalization

To use microarray datasets from the GEO database the studies can be downloaded directly executing R code. Therefore, initializing the directory and creating the required folders is needed before downloading GEO studies. The objects `targetcol` should also be entitled "target". To run this package analyzing DEGs and identifying diagnostic and prognostic gene signatures for the seperation of two groups (or subtypes, healthy vs cancer, etc) we here named `controlname` and `targetname` exemplary as "control" and "Lung Cancer".  

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
# initialize filePath:
filePath <- tempdir()
maindir <- "./geodata/"
datadir <- paste0(maindir, "data/")
dir.create(maindir)
dir.create(datadir)

# initialize more infos on the study
targetcol <- "target" # should be named "target"
controlname <- "Control"
targetname <- "Lung Cancer"
```

Insert the GEO accesion numbers of the desired GEO studies into the `getGEO()` function. This will download the datasets containing expressin data, pheno and feature data as well as the annotation of the probe sets and store it as .txt.gz files in the "data" folder. 

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
GSE18842 <- GEOquery::getGEO("GSE18842", destdir = datadir)
GSE19804 <- GEOquery::getGEO("GSE19804", destdir = datadir)

# extracting expressionSets
eset1 <- GSE18842[[1]]
eset2 <- GSE19804[[1]]
```

Morover, the approach enables to download raw data as CEL files, uncompress it and conduct GCRMA normalization. For this, further commands are necessary to subsequently import the CEL files into the R environment. Accordingly `esetC` represents a normalized epressionSet though only containing expression data without pheno and feature data. These data can be added later. (Don't be confused by our example. We could have downloaded the study GSE19188 directly by applying `getGEO()` instead of downlaoding the raw data and adding pheno and feature data afterwards. We just wanted to demonstrating you how to use raw data in form of CEL files)

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
# download raw data, uncompress them and execute GCRMA normalization
# actually not necessary for GSE19188 but to showcase feasibility of dealing with .CEL files
GEOquery::getGEOSuppFiles("GSE19188", baseDir = filePath)
utils::untar(paste0(filePath, "/GSE19188/GSE19188_RAW.tar"), exdir=maindir)
cels <- list.files(maindir, pattern = "[gz]")
tryCatch({
  sapply(paste(maindir, cels, sep="/"), GEOquery::gunzip)
  # at this point, first the textfile "phenodata.txt must be generated
  ## For downloading raw data use the function: getGEOSuppFiles(""). After downloading and uncompressing the raw file (*_RAW.tar) we need to capture
  ## the experimental information. Therefore a text file is created in the terminal window or rather in the command line.
  system("bash -c 'ls ./geodata/*.CEL > ./geodata/phenodata.txt'")
  ## The command: ls data/*.CEL > data/phenodata.txt puts all .CEl files in one text file.
  ## After this step the normalization can be started.
}, error = function(e){
  print(e)
})

cels <- list.files(maindir, pattern = "CEL$")

celfiles <- paste0(maindir, cels)
# use justGCRMA now due to memory issues with gcrma --> justGCRMA is much more memory efficient
# https://www.rdocumentation.org/packages/gcrma/versions/2.44.0/topics/justGCRMA
esetC <- gcrma::justGCRMA(filenames = celfiles, fast = TRUE)
# call garbage collection here
gc()

# old code
#celfiles <- affy::ReadAffy(filenames = gsub("\\.gz", "", cels), celfile.path = "./geodata")
#esetC <- gcrma::gcrma(celfiles, fast = TRUE)

# esetC = now is a normalized expressionSet, but without f- and pData
# the f- and pData are transfered manually from dataset GSE19188

GSE19188 <- GEOquery::getGEO("GSE19188", destdir = datadir)

eset3 <- GSE19188[[1]]
pData3 <- Biobase::pData(eset3)
fData3 <- Biobase::fData(eset3)
Biobase::pData(esetC) <- pData3
Biobase::fData(esetC) <- fData3
colnames(Biobase::exprs(esetC)) <- colnames(Biobase::exprs(eset3))
Biobase::annotation(esetC) <- Biobase::annotation(eset3)
eset3 <- esetC
```

## Formatting the phenoData 

The phenoData of the expresssionSets need to be formatted that the phenoData columns overlap perfectly. This means that all expressionSets should contain identical columns in the same order. Not consistent columns need to get removed. For a simpler procedure we recommend only retaining phenoData columns required for the downstream analysis (e.g. targetcol and survival data). 

```{r}
# create version b, that original expressionSets persist
eset1b <- eset1
eset2b <- eset2
eset3b <- eset3

# show number of samples and phenoData columns
dim(Biobase::pData(eset1b))
dim(Biobase::pData(eset2b))
dim(Biobase::pData(eset3b))
# depict phenoData column names
colnames(Biobase::pData(eset1b))
colnames(Biobase::pData(eset2b))
colnames(Biobase::pData(eset3b))


# remove unnecessary phenoData
# rename desired column names and if necessary the column entries (as we did for 'Tumor')

# eset1b
eset1b$relation <- NULL
eset1b$characteristics_ch1 <- NULL
eset1b$`sample type:ch1` <- NULL
eset1b$`tissue:ch1` <- NULL
colnames(Biobase::pData(eset1b))[8] = targetcol
levels(eset1b[[targetcol]]) = c(controlname, targetname)
dim(eset1b)


# eset2b
eset2b$characteristics_ch1.2 <- NULL
eset2b$characteristics_ch1.3 <- NULL
eset2b$contact_department <- NULL
eset2b$characteristics_ch1 <- NULL
eset2b$`age:ch1` <- NULL
eset2b$`gender:ch1` <- NULL
eset2b$`stage:ch1` <- NULL
eset2b$`tissue:ch1` <- NULL
colnames(Biobase::pData(eset2b))[8] = targetcol
levels(eset2b[[targetcol]]) = c(controlname, targetname)
dim(eset2b)


# eset3b
eset3b$characteristics_ch1.2 <- NULL
eset3b$characteristics_ch1.3 <- NULL
eset3b$characteristics_ch1.4 <- NULL
eset3b$contact_phone <- NULL
eset3b$contact_fax <- NULL
eset3b$contact_department <- NULL
eset3b$contact_web_link <- NULL
eset3b$relation <- NULL
eset3b$source_name_ch1 <- NULL
colnames(Biobase::pData(eset3b))[9] = targetcol
levels(eset3b[[targetcol]]) = c(controlname, targetname)
pNew <- Biobase::pData(eset3b)[,c(1:7,9,8,10:29)] # reorder columns
Biobase::pData(eset3b) = pNew
dim(eset3b)

# now, all phenoData columns are consistent
cbind(colnames(Biobase::pData(eset1b)), colnames(Biobase::pData(eset2b)), 
      colnames(Biobase::pData(eset3b)))
```


# Preparations for utilizing the sigident package 

These variables need to be defined for the package functions to work. One could use them also as arguments directly in the respective function. However, we think it is more clearly to define them here at the beginning and to refer at each function to the respective variable.  


```{r}
plotdir <- "./plots/"
dir.create(plotdir)
csvdir <- "./csv/"
dir.create(csvdir)
species <- "Hs"
OrgDb <- "org.Hs.eg.db"
organism <- "hsa"
pathwayid <- "hsa04110"
seed <- 111
split <- 0.8
```


## Extract Metadata 

TODO Description here.

### Extract sample metadata 

```{r}
# workaround
GSE18842_meta <- data.table::data.table(
  cbind(
    study = "GSE18842",
    sample = eset1b@phenoData@data[,"geo_accession"],
    target = as.character(eset1b@phenoData@data[,targetcol])
  )
)

GSE19804_meta <- data.table::data.table(
  cbind(
    study = "GSE19804",
    sample = eset2b@phenoData@data[,"geo_accession"],
    target = as.character(eset2b@phenoData@data[,targetcol])
  )
)

GSE19188_meta <- data.table::data.table(
  cbind(
    study = "GSE19188",
    sample = eset3b@phenoData@data[,"geo_accession"],
    target = as.character(eset3b@phenoData@data[,targetcol])
  )
)

sampleMetadata <- rbind(GSE18842_meta,
                        GSE19804_meta,
                        GSE19188_meta)
```

### Define study metadata 

```{r}
studyMetadata <- data.frame(
  cbind(study = c("GSE18842", "GSE19804", "GSE19188", "GSE30219"),
        discovery = as.logical(c(1, 1, 1, 0)),
        validation = as.logical(c(0, 0, 0, 1))
  ),
  stringsAsFactors = FALSE
)
```

## Merging and Batch Correction

TODO description here

First, a boxplot is created with the included samples on the x-axis and the standardised expression values on the y-axis. `mergedset` results as output from above described merging and represents a matrix containing the genes (Entrez ID) as rows and the samples as columns.

```{r}
#-----------------------------------------#
##### 4. Merging and Batch Correction #####
#-----------------------------------------#


esets=c(eset1b,eset2b,eset3b)
mergedset <- sigident::merge_(esets)
# merging 3 esets, resulting in a mergeset containing 367 samples and 54675 genes
```


```{r}
# visualize log2 transformed expression values of the merged data set
filename <- paste0(plotdir, "boxplot_merged_data.jpeg")
jpeg(filename, width = 2000, height = 1000, res = 150, units = "px")
boxplot(mergedset@assayData$exprs, main = "Merged data before batch correction", 
        xlab = "Samples", ylab ="Expression value")
dev.off()
```

```{r out.width='80%'}
knitr::include_graphics(filename)
```

### Discovering Batch Effects 

```{r message=FALSE, warning=FALSE, error=FALSE}
# generate vector with column diagnosis
table(mergedset@phenoData@data[[targetcol]])

dd <- sigident::createDiagnosisDesign_(sampleMetadata = sampleMetadata,
                                       studyMetadata = studyMetadata,
                                       controlname = controlname,
                                       targetname = targetname,
                                       targetcol = targetcol)
diagnosis <- dd$diagnosis
length(diagnosis)
table(diagnosis)

design <- dd$design

# determine batches with number of samples
batch <- sigident::createBatch_(studyMetadata = studyMetadata,
                                sampleMetadata = sampleMetadata)
# check batch assignment
table(batch)
length(batch)

# Batch Effect Correction 
# removing batch effects computing linear models, take variance between the diagnosis into consideration
mergeset <- sigident::createCombat_(mergedset = mergedset, batch = batch, design = design)
```

`mergeset` results as output from above described merging and represents a matrix containing the genes (Entrez ID) as rows and the samples as columns.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
DF <- mergedset@assayData$exprs
gPCA_before <- sigident::batchDetection_(mergeset = DF,
                                          batch = batch)
filename <- paste0(plotdir, "PCplot_before.png")
sigident::createBatchPlot_(correction_obj = gPCA_before,
                           filename = filename,
                           time = "before")
```

```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "import_boxplot.png")
sigident::createImportBoxplot_(mergeset, filename)
```

```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide'}
# remove unneeded objects
rm("DF", "GSE18842_meta", "GSE19188_meta", "GSE19804_meta")
```


### Visualize batch effects  

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
gPCA_after <- sigident::batchDetection_(mergeset = mergeset,
                                        batch = batch)
filename <- paste0(plotdir, "PCplot_after.png")
sigident::createBatchPlot_(correction_obj = gPCA_after,
                           filename = filename,
                           time = "after")
```

```{r out.width='80%'}
knitr::include_graphics(filename)
```

# DEG Analysis 

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
q_selection <- NULL # use default setting
deg_q <- sigident::qSelection_(sampleMetadata = sampleMetadata,
                               deg.q.selection = q_selection)

genes <- sigident::identifyDEGs_(mergeset = mergeset,
                                 design = design,
                                 qValue = deg_q)

# heatmap creation
filename <- paste0(plotdir, "DEG_heatmap.png")
# create colors for map
ht_colors <- sigident::colorHeatmap_(sampleMetadata = sampleMetadata,
                                     studyMetadata = studyMetadata,
                                     targetcol = targetcol,
                                     controlname = controlname) # cancer = red
sigident::createDEGheatmap_(mergeset = mergeset,
                            genes = genes,
                            patientcolors = ht_colors,
                            filename = filename)
```

```{r out.width='80%'}
knitr::include_graphics(filename)
```

TODO description here

```{r}
deg_info <- sigident::exportDEGannotations_(mergedset = mergedset)
data.table::fwrite(deg_info, paste0(csvdir, "DEG_info.csv"))
```

```{r}
knitr::kable(head(deg_info))
```

```{r}
deg_results <- sigident::limmaTopTable_(mergeset = mergeset,
                                        design = design,
                                        qValue = deg_q)
data.table::fwrite(deg_results, paste0(csvdir, "DEG_results.csv"))
```

```{r}
knitr::kable(head(deg_results))
```

# Gene enrichment 

TODO Description here.

```{r}
deg_entrez <- unique(genes)
```


## Test for over-representation 

TODO Description here.

```{r}
enr_topgo <- sigident::extractGOterms_(entrez = deg_entrez,
                                       species = species)
```
```{r}
dim(enr_topgo)
knitr::kable(head(enr_topgo))
```

TODO Description here.

```{r}
enr_topkegg <- sigident::extractKEGGterms_(entrez = deg_entrez,
                                           species = species)
```
```{r}
dim(enr_topkegg)
knitr::kable(head(enr_topkegg))
```

TODO Description here.

```{r}
enr_fitlm <- sigident::goDiffReg_(mergeset = mergeset,
                                  design = design)

enr_fitlm_topgo <- sigident::extractGOterms_(entrez = enr_fitlm,
                                             species = species,
                                             FDR = 0.01)
data.table::fwrite(enr_fitlm_topgo, paste0(csvdir, "Top_GO_fitlm.csv"))
```
```{r}
dim(enr_fitlm_topgo)
knitr::kable(head(enr_fitlm_topgo))
```

```{r}
enr_fitlm_topkegg <- sigident::extractKEGGterms_(entrez = enr_fitlm,
                                                 species = species)
data.table::fwrite(enr_fitlm_topkegg, paste0(csvdir, "Top_KEGG_fitlm.csv"))
```
```{r}
dim(enr_fitlm_topkegg)
knitr::kable(head(enr_fitlm_topkegg))
```

## GO Analysis

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
enr_analysis <- sigident::goEnrichmentAnalysis_(entrez = deg_entrez,
                                                OrgDB = OrgDb,
                                                organism = organism,
                                                fitlm = enr_fitlm,
                                                pathwayid = pathwayid,
                                                species = organism,
                                                plotdir = plotdir)
```
```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "/", organism, "04110.png")
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```
```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "/", organism, "04110.pathview.png")
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "Enriched_GO.png")
sigident::createEnrichtedBarplot_(enrichmentobj = enr_analysis$go,
                                  type = "GO",
                                  filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "Enriched_KEGG.png")
sigident::createEnrichtedBarplot_(enrichmentobj = enr_analysis$kegg,
                                  type = "KEGG",
                                  filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```


# Identification of diagnostic signatures 

TODO Description here.

## Split data into training and test dataset

TODO Description here.

```{r}
training_list <- sigident::createTrainingTest_(diagnosis = diagnosis,
                                               mergeset = mergeset,
                                               split = split,
                                               seed = seed)
```

## Lasso regression

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
diagnostic_lasso <- sigident::signature_(traininglist = training_list,
                                         type = "lasso",
                                         nfolds = 10,
                                         seed = seed)
filename <- paste0(plotdir, "CV_lasso.png")
sigident::createCVPlot_(cv_obj = diagnostic_lasso$fitCV,
                        filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "ROC_Lasso.min.png")
sigident::createROCplot_(roc = diagnostic_lasso$roc.min,
                         filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "ROC_Lasso.1se.png")
sigident::createROCplot_(roc = diagnostic_lasso$roc.1se,
                         filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```


## Elastic net regression 

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
diagnostic_elasticnet <- sigident::signature_(traininglist = training_list,
                                              type = "elastic",
                                              alpha = 0.9,
                                              nfolds = 10,
                                              seed = seed)
filename <- paste0(plotdir, "CV_elasticNet.png")
sigident::createCVPlot_(cv_obj = diagnostic_elasticnet$fitCV,
                        filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "ROC_elasticNet.min.png")
sigident::createROCplot_(roc = diagnostic_elasticnet$roc.min,
                         filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "ROC_elasticNet.1se.png")
sigident::createROCplot_(roc = diagnostic_elasticnet$roc.1se,
                         filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```


## Gridsearch to find alpha and lambda

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
diagnostic_glmGrid <- sigident::signature_(traininglist = training_list,
                                           type = "grid",
                                           nfolds = 10,
                                           seed = seed)
# plot model of gridsearch
filename <- paste0(plotdir, "Gridsearch_model.png")
sigident::createGridModelPlot_(model = diagnostic_glmGrid$caret.train,
                               filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
# plot variable importance of gridsearch
filename <- paste0(plotdir, "Gridsearch_variable_importance.png")
sigident::createGridVarImpPlot_(model = diagnostic_glmGrid$caret.train,
                                filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
# create roc plot
filename <- paste0(plotdir, "ROC_elasticNet.grid.png")
sigident::createROCplot_(roc = diagnostic_glmGrid$roc.elasticNet,
                         filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

# Analysis of Diagnostic Models 

## Comparing Model Performance  

These models can be included into one big list and the results can then be printed using the function `compareDiagnosticModels()`.

```{r}
diagnosticModels <- list(
  lasso.min = list(model = diagnostic_lasso$lambda.min,
                   confmat = diagnostic_lasso$confmat.min,
                   auc = as.numeric(diagnostic_lasso$roc.min$auc)),
  lasso.1se = list(model = diagnostic_lasso$lambda.1se,
                   confmat = diagnostic_lasso$confmat.1se,
                   auc = as.numeric(diagnostic_lasso$roc.1se$auc)),
  elastic.min = list(model = diagnostic_elasticnet$lambda.min,
                     confmat = diagnostic_elasticnet$confmat.min,
                     auc = as.numeric(diagnostic_elasticnet$roc.min$auc)),
  elastic.1se = list(model = diagnostic_elasticnet$lambda.1se,
                     confmat = diagnostic_elasticnet$confmat.1se,
                     auc = as.numeric(diagnostic_elasticnet$roc.1se$auc)),
  elastic.grid = list(model = diagnostic_glmGrid$elasticNet.auto,
                      confmat = diagnostic_glmGrid$confmat.elasticNet,
                      auc = as.numeric(diagnostic_glmGrid$roc.elasticNet$auc))
  )
```

```{r}
knitr::kable(
  sigident::compareDiagnosticModels(diagnosticModels)
)
```

## Export selected Entrez-IDs

```{r}
# map selected variables and export 
signature_genes <- sigident::geneMapSig_(mergeset = mergeset, model = diagnostic_lasso$lambda.min)
data.table::fwrite(signature_genes, paste0(csvdir, "signature_genes_model1.csv"))
```

# Identification of Prognostic Signature

## Create a list that contains specifications of the study/studies that contain(s) survival time information 

```{r}
discoverystudies.w.timedata <- list("GSE19188" = list(timecol = "characteristics_ch1.2",
                                                      statuscol = "characteristics_ch1.3",
                                                      targetcolname = "characteristics_ch1"))
```

```{r}
#---------------------------------------------------#
##### 8. Identification of Prognostic Signature #####
#---------------------------------------------------#

### 8.A Identification of survival correlated genes
survivalDataTable <- getSurvivalTime_(studyMetadata = studyMetadata,
                                      sampleMetadata = sampleMetadata,
                                      discoverystudies.w.timedata = discoverystudies.w.timedata,
                                      targetname = targetname,
                                      controlname = controlname,
                                      targetcol = targetcol,
                                      datadir = datadir)
```


# An alternative workflow: one generic function 

In order to simplify the whole abovedescribed workflow, we wrapped all these functions into one big function.

## Preprocessing: 

The preprocessing needs to be conform with the above described approach.

## Run `sigidentMicroarray`-Function 


```{r eval = FALSE}
diagnosticModels <- sigident::sigidentMicroarray(mergeset = mergeset,
                                                 controlname = "Control",
                                                 targetname = "Lung Cancer",
                                                 studyMetadata = studyMetadata,
                                                 sampleMetadata = sampleMetadata,
                                                 species = "Hs",
                                                 OrgDB = "org.Hs.eg.db",
                                                 organism = "hsa",
                                                 pathwayid = "hsa04110",
                                                 deg.q.selection = NULL,
                                                 seed = 111,
                                                 nfolds = 10,
                                                 split = 0.8,
                                                 plotdir = "./plots/",
                                                 csvdir = "./tables/",
                                                 targetcol = "target")
```

