---
title: "sigident - Howto Microarray"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{sigident-Howto_Microarray}
  %\VignetteEngine{knitr::knitr}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(sigident)
library(knitr)
library(mergeGEO)
```

# Prerequisites and Installation 

## Install mergeGEO package

```{r eval=FALSE}
devtools::install_git("https://gitlab.miracum.org/clearly/mergegeo.git")
```

## Install sigident package

```{r eval=FALSE}
devtools::install_git("https://gitlab.miracum.org/clearly/sigident.git")
```

# Data Preparation 

In order to use this R package and its functions, you need to prepare a merged gene expression dataset. 
This example uses the GEO lung cancer studies "GSE18842", "GSE19804" and "GSE19188", which contain in total 367 samples (197 tumor; 170 non-tumor) and 54,675 transcripts gained by using Affymetrix GeneChip Human Genome U133 Plus 2.0 Array (platform GPL570).

We therefore created the R package 'mergeGEO' that organizes a method to merge GEO datasets, originally described and published by Hughey & Butte (2015) in one R function.
For a detailled background, please refer to the original publication ([https://doi.org/10.1093/nar/gkv229](https://doi.org/10.1093/nar/gkv229)) and its supplemental R files (licensed under GPLv3, [https://doi.org/10.5281/zenodo.16006](https://doi.org/10.5281/zenodo.16006)). 

To get a detailled background on how to use 'mergeGEO', please visit its package vignette: ([link](link))

## Download of GEO datasets using mergeGEO 

During the utilization of mergeGEO, the GEO datasets are downloaded, normalized and batch corrected. In order to provide a working example, we included exemplary files containing the required study metadata and the sample metadata.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
studymetadata = "lungcancer_study_metadata.csv"
samplemetadata = "lungcancer_sample_metadata.csv"
studyname <- "lungcancer"
denovo <- TRUE
metadatadir <- "./metadata/"
dir.create(metadatadir)

# copy prepared files to metadatadir
file.copy(from=system.file("./demofiles/lungcancer_study_metadata.csv",
                           package = "mergeGEO"),
          to="./metadata/lungcancer_study_metadata.csv")
file.copy(from=system.file("./demofiles/lungcancer_sample_metadata.csv",
                           package = "mergeGEO"),
          to="./metadata/lungcancer_sample_metadata.csv")

mergeset <- mergeGEO::mergeGEO(studymetadata = studymetadata,
                               samplemetadata = samplemetadata,
                               studyname = studyname,
                               denovo = denovo,
                               metadatadir = metadatadir)
```


## Loading Metadata 

TODO Description here.

```{r}
studyMetadata <- mergeGEO::readStudyMetadata_(studymetadataFilename = paste0(metadatadir, studymetadata))
sampleMetadata <- mergeGEO::readSampleMetadata_(samplemetadataFilename = paste0(metadatadir, samplemetadata),
                                             studyMetadata = studyMetadata)
```

# Preparations for utilizing the sigident package 

These variables need to be defined for the package functions to work. One could use them also as arguments directly in the respective function. However, we think it is more clearly to define them here at the beginning and to refer at each function to the respective variable.  


```{r}
plotdir <- "./plots/"
dir.create(plotdir)
csvdir <- "./csv/"
dir.create(csvdir)
targetcol <- "target"
controlname <- "Control"
targetname <- "Lung Cancer"
species <- "Hs"
OrgDb <- "org.Hs.eg.db"
organism <- "hsa"
pathwayid <- "hsa04110"
seed <- 111
traintest.split <- 0.8
```

First, a boxplot is created with the included samples on the x-axis and the standardised expression values on the y-axis. `mergeset` results as output from the function `mergeGEO()` and represents a matrix containing the genes (Entrez ID) as rows and the samples as columns.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "import_boxplot.png")
createImportBoxplot_(mergeset, filename)
```

```{r out.width='80%'}
knitr::include_graphics(filename)
```


# Create experiment desciption and visualize batch effects  

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
dd <- createDiagnosisDesign_(sampleMetadata = sampleMetadata,
                             studyMetadata = studyMetadata,
                             controlname = controlname,
                             targetname = targetname,
                             targetcol = targetcol)
diagnosis <- dd$diagnosis
design <- dd$design

batch <- createBatch_(studyMetadata = studyMetadata,
                      sampleMetadata = sampleMetadata)

gPCA_after <- batchCorrection_(mergeset = mergeset,
                               batch = batch)
filename <- paste0(plotdir, "PCplot_after.png")
createBatchPlot_(correction_obj = gPCA_after,
                 filename = filename,
                 time = "after")
```

```{r out.width='80%'}
knitr::include_graphics(filename)
```

# DEG Analysis 

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
q_selection <- NULL # use default setting
deg_q <- qSelection_(sampleMetadata = sampleMetadata,
                     deg.q.selection = q_selection)

genes <- identify.DEGs_(mergeset = mergeset,
                        design = design,
                        qValue = deg_q)

# heatmap creation
filename <- paste0(plotdir, "DEG_heatmap.png")
# create colors for map
ht_colors <- colorHeatmap_(sampleMetadata = sampleMetadata,
                           targetcol = targetcol,
                           controlname = controlname) # cancer = red
createDEGheatmap_(mergeset = mergeset,
                  genes = genes,
                  patientcolors = ht_colors,
                  filename = filename)
```

```{r out.width='80%'}
knitr::include_graphics(filename)
```

<!-- TODO start here 13.9. Lorenz -->

<!-- ```{r} -->
<!-- deg_info <- exportDEGannotations_(mergeset = mergeset, -->
<!--                                   genes = genes) -->
<!-- data.table::fwrite(deg_info, paste0(csvdir, "DEG_info.csv")) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- knitr::kable(head(deg_info)) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- deg_results <- limmaTopTable_(BatchRemovedExprs = combat, -->
<!--                               design = design, -->
<!--                               qValue = deg_q) -->
<!-- data.table::fwrite(deg_results, paste0(csvdir, "DEG_results.csv")) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- knitr::kable(head(deg_results)) -->
<!-- ``` -->

# Gene enrichment 

TODO Description here.

```{r}
deg_entrez <- unique(genes)
```


## Test for over-representation 

TODO Description here.

```{r}
enr_topgo <- extractGOterms_(entrez = deg_entrez,
                             species = species)
```
```{r}
dim(enr_topgo)
knitr::kable(head(enr_topgo))
```

TODO Description here.

```{r}
enr_topkegg <- extractKEGGterms_(entrez = deg_entrez,
                                 species = species)
```
```{r}
dim(enr_topkegg)
knitr::kable(head(enr_topkegg))
```

TODO Description here.

```{r}
enr_fitlm <- goDiffReg_(mergeset = mergeset,
                        design = design)

enr_fitlm_topgo <- extractGOterms_(entrez = enr_fitlm,
                                   species = species,
                                   FDR = 0.01)
data.table::fwrite(enr_fitlm_topgo, paste0(csvdir, "Top_GO_fitlm.csv"))
```
```{r}
dim(enr_fitlm_topgo)
knitr::kable(head(enr_fitlm_topgo))
```

```{r}
enr_fitlm_topkegg <- extractKEGGterms_(entrez = enr_fitlm,
                                       species = species)
data.table::fwrite(enr_fitlm_topkegg, paste0(csvdir, "Top_KEGG_fitlm.csv"))
```
```{r}
dim(enr_fitlm_topkegg)
knitr::kable(head(enr_fitlm_topkegg))
```

## GO Analysis

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
enr_analysis <- goEnrichmentAnalysis_(entrez = deg_entrez,
                                      OrgDB = OrgDb,
                                      organism = organism,
                                      fitlm = enr_fitlm,
                                      pathwayid = pathwayid,
                                      species = organism,
                                      plotdir = plotdir)
```
```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "/", organism, "04110.png")
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```
```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "/", organism, "04110.pathview.png")
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "Enriched_GO.png")
createEnrichtedBarplot_(enrichmentobj = enr_analysis$go,
                        type = "GO",
                        filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "Enriched_KEGG.png")
createEnrichtedBarplot_(enrichmentobj = enr_analysis$kegg,
                        type = "KEGG",
                        filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```


# Identification of diagnostic signatures 

TODO Description here.

## Split data into training and test dataset

TODO Description here.

```{r}
training_list <- createTrainingTest_(diagnosis = diagnosis,
                                     mergeset = mergeset,
                                     split = traintest.split,
                                     seed = seed)
```

## Lasso regression

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
diagnostic_lasso <- glmnetSignature_(traininglist = training_list,
                                     type = "lasso",
                                     nfolds = 10,
                                     seed = seed)
filename <- paste0(plotdir, "CV_lasso.png")
createCVPlot_(cv_obj = diagnostic_lasso$fitCV,
              filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "ROC_Lasso.min.png")
createROCplot_(roc = diagnostic_lasso$roc.min,
               filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "ROC_Lasso.1se.png")
createROCplot_(roc = diagnostic_lasso$roc.1se,
               filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```


## Elastic net regression 

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
diagnostic_elasticnet <- glmnetSignature_(traininglist = training_list,
                                          type = "elastic",
                                          alpha = 0.9,
                                          nfolds = 10,
                                          seed = seed)
filename <- paste0(plotdir, "CV_elasticNet.png")
createCVPlot_(cv_obj = diagnostic_elasticnet$fitCV,
              filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "ROC_elasticNet.min.png")
createROCplot_(roc = diagnostic_elasticnet$roc.min,
               filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
filename <- paste0(plotdir, "ROC_elasticNet.1se.png")
createROCplot_(roc = diagnostic_elasticnet$roc.1se,
               filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```


## Gridsearch to find alpha and lambda

TODO Description here.

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
diagnostic_glmGrid <- glmnetSignature_(traininglist = training_list,
                                       type = "grid",
                                       nfolds = 10,
                                       seed = seed)
# plot model of gridsearch
filename <- paste0(plotdir, "Gridsearch_model.png")
createGridModelPlot_(model = diagnostic_glmGrid$caret.train,
                     filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
# plot variable importance of gridsearch
filename <- paste0(plotdir, "Gridsearch_variable_importance.png")
createGridVarImpPlot_(model = diagnostic_glmGrid$caret.train,
                      filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```

```{r results='hide', message=FALSE, warning=FALSE, error=FALSE}
# create roc plot
filename <- paste0(plotdir, "ROC_elasticNet.grid.png")
createROCplot_(roc = diagnostic_glmGrid$roc.elasticNet,
               filename = filename)
```
```{r out.width='80%'}
knitr::include_graphics(filename)
```




# An alternative workflow: one generic function 

In order to simplify the whole abovedescribed workflow, we wrapped all these functions into one big function.

## Preprocessing: 

TODO Description here.

### Merge data sets  

TODO Description here.

```{r eval = FALSE}
studymetadata = "lungcancer_study_metadata.csv"
samplemetadata = "lungcancer_sample_metadata.csv"
studyname <- "lungcancer"
denovo <- TRUE
metadatadir <- "./metadata/"
dir.create(metadatadir)

# copy prepared files to metadatadir
file.copy(from=system.file("./demofiles/lungcancer_study_metadata.csv",
                           package = "mergeGEO"),
          to="./metadata/lungcancer_study_metadata.csv")
file.copy(from=system.file("./demofiles/lungcancer_sample_metadata.csv",
                           package = "mergeGEO"),
          to="./metadata/lungcancer_sample_metadata.csv")

mergeset <- mergeGEO(studymetadata = studymetadata,
                     samplemetadata = samplemetadata,
                     studyname = studyname,
                     denovo = denovo,
                     metadatadir = metadatadir)
```

### Read metadata 

TODO Description here.

```{r eval = FALSE}
studyMetadata <- mergeGEO::readStudyMetadata_(studymetadataFilename = paste0(metadatadir,
                                                                             studymetadata))
sampleMetadata <- mergeGEO::readSampleMetadata_(samplemetadataFilename = paste0(metadatadir,
                                                                                samplemetadata),
                                                studyMetadata = studyMetadata)
```

## Run `sigidentMicroarray`-Function 

TODO Description here.

```{r eval = FALSE}
diagnosticModels <- sigidentMicroarray(mergeset = mergeset,
                                       controlname = "Control",
                                       targetname = "Lung Cancer",
                                       studyMetadata = studyMetadata,
                                       sampleMetadata = sampleMetadata,
                                       species = "Hs",
                                       OrgDb = "org.Hs.eg.db",
                                       organism = "hsa",
                                       pathwayid = "hsa04110",
                                       deg.q.selection = NULL,
                                       seed = 111,
                                       nfolds = 10,
                                       traintest.split = 0.8,
                                       plotdir = "./plots/",
                                       csvdir = "./tables/",
                                       targetcol = "target"))
```

